### PR TITLE
Move apiclient_util to apiclient/util package and add unit tests

### PR DESCRIPTION
**Pull request by rifung**
Moving to util package was necessary to prevent a circular dependency
when using storage/mem in the tests

### CODE REVIEW

**Comment by MHBauer** [comment]
is this the same as #293 but after a rebase?

**Comment by rifung** [comment]
@MHBauer Yes it is. I am now realizing I probably should have tried to do a merge instead of a rebase because that would make it so I can still push to the same remote branch on my forked repo.

Oops!

**Comment by MHBauer** [comment]
I'll let someone else more familiar with the code formation to comment on the correct location. Maybe moving to a util pkg is appropriate here.

My comment is related to the necessity. This is where you would use test package. That is, instead of `apiclient_util_test.go` being `package apiclient`, it would be `package apiclient_test`. There would be no cyclic import.

**Comment by MHBauer** [comment]
also, golint is unhappy.

**Comment by MHBauer** [comment]
This is also the one time using a dot import is okay, to import the original non-test package.

`. "github.com/kubernetes-incubator/service-catalog/pkg/controller/apiclient"`

**Comment by rifung** [comment]
Thanks for the suggestions. As you can see I need to get more familiar with Go. I will update the patch and include the changes. I am going to assume that moving to util isn't necessary here due to the fact nobody decided to move it before.

**Comment by MHBauer** [comment]
@bmelville what do you think?

File: pkg/controller/apiclient/apiclient_util_test.go [hunk 1]
```diff
@@ -14,18 +14,180 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package apiclient
+package apiclient_test
```
Comment (arschles) [inline pkg/controller/apiclient/apiclient_util_test.go]: can you leave this in the `apiclient` package, and remove the `. "github.com/kubernetes-incubator/service-catalog/pkg/controller/apiclient"` import below. as-is, there's no need to put these tests in a different package

Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: I'm not sure if you read the previous comments but the reason I put it into the apiclient_test package was to prevent a circular dependency since this relies on apiclient/mem but apiclient/mem uses apiclient, thus we can't put this into apiclient (as it is).

If you have another idea on how we should structure the packages or a way to get around this though I'd love to hear it.

File: pkg/controller/apiclient/apiclient_util_test.go [hunk 2]
```diff
@@ -14,18 +14,180 @@ See the License for the specific language governing permissions and
...
+		svcClss[0].Plans[0].Name); err == nil {
+
+		t.Error("Expected err when using wrong service name")
+	}
+}
+
+// test service classes with multiple plans
+func TestGetServicePlanInfoMany(t *testing.T) {
+	apiclient := mem.NewAPIClient()
+	svcClss := createServiceClasses(3, 3)
```
Comment (pmorie) [inline pkg/controller/apiclient/apiclient_util_test.go]: I find this line to be fairly hard to read - it's not clear at all what the arguments are.

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 1 above)
Comment (pmorie) [inline pkg/controller/apiclient/apiclient_util_test.go]: I'm agnostic on how you guys choose to solve this problem, but I would like @rifung to be looking at the k8s reaction chain fakes.

**Comment by arschles** [comment]
Duh, sorry @rifung. I'll admit to not looking at this PR with my full, undivided attention. Consider my previous comment resolved for now and feel free to merge if all other comments are resolved.

I'll look when I can for a reasonable solution between now and reaction chain fakes (I am fine with going that route, since it's pretty standard for k8s-land), if there is one.

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 2 above)
Comment (MHBauer) [inline pkg/controller/apiclient/apiclient_util_test.go]: create service classes with service plans? Something like that?

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 1 above)
Comment (krancour) [inline pkg/controller/apiclient/apiclient_util_test.go]: Humbly, I think both views on how to resolve the would-be dependency cycle here are wrong. If you look at these tests, they are very specifically testing the in-memory implementation of the APIClient interface. That means these tests belong alongside that implementation _in_ the `pkg/controller/apiclient/mem` package.

Also, although some tests similar to this do already exist and aren't completely useless, more tests of this ilk might not give us the biggest bang for the buck. The in-memory implementation of this interface is, essentially, a fake implementation used only to facilitate tests. These are tests for test code. As I said-- that's not completely without value, but as @pmorie has indicated, longer term we want to stop using this implementation and use the reactor chain pattern here instead. (Which will courtesy of some code generation that I don't claim to understand.)

Bottom line-- I would suggest that we move this to where it belongs and not fuss over it too much further since its value is limited.

Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: @krancour Ah that's a good point; I'll move this to the mem package.

Can someone please direct me to the reaction chain fakes? I can't seem to find anything about it..

Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: @krancour actually I take that back. This isn't intended to test the in memory implementation of APIClient but to test apiclient_util, which is in the apiclient package.

That being said, moving this to the mem package would still work but doesn't seem like the right thing to do..

That's why I moved both apiclient_util and the test to its own util package, but someone made the suggestion to leave that here and instead make a new test package instead.

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 2 above)
Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: @pmorie @MHBauer this creates 3 services classes each with 3 service plans.

I would agree it's not obvious at first what it does, but I'm also not too sure how to improve it. I can add a comment above the function definition, but perhaps you have a better idea for naming or something?

Comment (MHBauer) [inline pkg/controller/apiclient/apiclient_util_test.go]: That was my suggestion:
`createServiceClassesWithServicePlans`
Some people don't like long names, so I can understand not liking it.

Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: Ohhh I'm sorry I didn't understand what you meant! I thought you were replying to the previous comment explaining what it was.

I've made the change you suggested (I am personally perfectly ok with long names)

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 1 above)
Comment (arschles) [inline pkg/controller/apiclient/apiclient_util_test.go]: @rifung got it. we don't have a generated client right now, so I don't think it's worth starting to build our own reaction chain code right now. let's wait until we can use the generated code to do it. this is fine for now

File: pkg/controller/apiclient/apiclient_util_test.go [hunk 3]
```diff
@@ -14,18 +14,180 @@ See the License for the specific language governing permissions and
...
+	apiclient.ServiceClasses().Create(&svcClss[0])
+
+	// test that getting with empty plan name works
+	if err := testGetServicePlanInfo(apiclient, &svcClss[0], &svcClss[0].Plans[0], ""); err != nil {
+		t.Error(err)
+	}
+
+	// test that getting with correct plan name works
+	if err := testGetServicePlanInfo(apiclient, &svcClss[0], &svcClss[0].Plans[0],
+		svcClss[0].Plans[0].Name); err != nil {
```
Comment (bmelville) [inline pkg/controller/apiclient/apiclient_util_test.go]: Just a general comment about line breaking throughout this. Go doesn't have a column restriction so this and others can be one line

File: pkg/controller/apiclient/apiclient_util_test.go [hunk 4]
```diff
@@ -14,18 +14,180 @@ See the License for the specific language governing permissions and
...
+
```
Comment (bmelville) [inline pkg/controller/apiclient/apiclient_util_test.go]: and can drop the newline here and elsewhere

File: pkg/controller/apiclient/apiclient_util_test.go [hunk 5]
```diff
@@ -14,18 +14,180 @@ See the License for the specific language governing permissions and
...
+		ret[i] = servicecatalog.ServicePlan{
+			Name:    "planName" + string(i+startCount),
+			OSBGUID: "planGUID" + string(i+startCount),
+		}
+	}
+	return ret
 }
 
-func TestGetBindingsForService(t *testing.T) {
+func testGetServicePlanInfo(apiclient APIClient,
```
Comment (bmelville) [inline pkg/controller/apiclient/apiclient_util_test.go]: Probably easier to read if you break this one up with one parameter per line

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 1 above)
Comment (bmelville) [inline pkg/controller/apiclient/apiclient_util_test.go]: seems like maybe the sane thing to do would be to have proper go package for apiclient utils in apiclient/util.

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 3 above)
Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: Done

File: pkg/controller/apiclient/apiclient_util_test.go (hunk 5 above)
Comment (rifung) [inline pkg/controller/apiclient/apiclient_util_test.go]: Done

**Comment by duglin** [comment]
@pmorie gives his LGTM via slack.
